### ISSUE TITLE
New info hints to enable network hardware atomics in RMA atomics

### ISSUE THREAD

**Issue by minsii** [mpiwg-rma/rma-issues#8]
# Problem

With proposal in [#114](https://github.com/mpi-forum/mpi-issues/issues/114), the MPI implementation can know which atomic operations are expected in the user program thus those operations may be safely offloaded to hardware if all of them are supported by the underlying network.

However, the active message based approach might be more beneficial for some of those operations if the network issuing cost is too expensive. For example:
- If the target datatype is non-contiguous and the data structure is not supported by the network hardware, the data may have to be divided into multiple contiguous chunks and associated with multiple atomic operations;
- Similarly, if the data size is so large that cannot be carried by single network atomic operation, multiple atomic operations have to be generated. 

To ensure atomicity, unfortunately, the implementation has to either always use network hardware atomics for all supported operations regardless of the overhead, or always use active message routines in case some operations with complex (or large) data might be issued. None of the above approaches is ideal for performance.

# Proposal

This proposal introduces two additional info keys:

- **accumulate_derived_dtype** - if set to false, then the implementation may assume that any accumulate calls for which atomicity is expected will not use derived datatypes. The default is true.

- **accumulate_max_bytes** - the info value can be any non-negative integer. It specifies the maximum data size to be used in an accumulate routine for which atomicity is expected. The default is unbounded.

With the above hints, the implementation can expect whether other processes' accumulate operations are handled by the network hardware, e.g., the operation is issued as a single network atomic operation if **accumulate_derived_dtype==true** and **accumulate_max_bytes** is smaller than the network atomics' length limitation. Thus, the local process can safely utilize network hardware atomics when possible.

# Impact on Implementations

The additional info keys gives more room for performance optimization, which can be used by implementations, if desired.

# Impact on Users

We do not expect any impact on existing applications. The atomicity of accumulate operations are always guaranteed with or without the proposed info keys.

**Comment by pavanbalaji** [comment]
Can you change "concurrent accumulate calls" to "concurrent accumulate calls for which atomicity is expected"?  Similarly, add equivalent wording to the max_bytes info key too?

**Comment by minsii** [comment]
Updated the info keys' description. Better now? 

I also removed "to the same target address location" because the user can use the hints to describe the operations only for entire window. For example, although we can somehow still enable HW atomics only for accumulates to location y in the following case, there is no way to properly describe it through the hints.
```
P0:                                                   P1:
accumulate(vector, x, P1);
accumulate(int, y, P1);                   accumulate(int, y, P1);
```

**Comment by pavanbalaji** [comment]
For the max_bytes description, the window size can be different for each target.  I think it might be better to not talk about window size at all and say that the default size is unbounded.

**Comment by pavanbalaji** [comment]
> Updated the info keys' description. Better now?
> 
> I also removed "to the same target address location" because the user can use the hints to describe the operations only for entire window. For example, although we can somehow still enable HW atomics only for accumulates to location y in the following case, there is no way to properly describe it through the hints.
> 
> ```
> P0:                                                   P1:
> accumulate(vector, x, P1);
> accumulate(int, y, P1);                   accumulate(int, y, P1);
> ```

Why is this a problem?  I would simply say no derived datatypes because I don't expect atomicity for the derived datatype part (this case was the exact reason for my first comment).

**Comment by minsii** [comment]
I see. Also updated `max_bytes` description.

**Comment by pavanbalaji** [comment]
Thanks.  Looks good to me!

**Comment by pavanbalaji** [comment]
Oh, sorry, one more comment: does the name `accumulate_derived_dtype` need to be updated or is it clear enough?

**Comment by minsii** [comment]
I changed "concurrent accumulate calls" to "any accumulate calls for which atomicity is expected", as the `for which atomicity is expected` part already excludes the case where only single operation updates the location (no atomicity needed).

**Comment by pavanbalaji** [comment]
I agree that the text is clear.  I was asking about the name `accumulate_derived_dtype`.

**Comment by minsii** [comment]
Ah, I see. Maybe `no_accumulate_derived_dtype` ?

**Comment by jeffhammond** [comment]
Nobody should ever expect atomicity for derived datatypes.  Or are we making a huge change to RMA semantics in MPI-4?

**Comment by pavanbalaji** [comment]
In all of these comments, we are only referring to element-wise atomicity, which is already guaranteed in MPI-3.1

**Comment by minsii** [comment]
@pavanbalaji : I guess all the accumulate hints use the term "accumulate" as atomic operation, but not as MPI accumulate call. E.g., `which_accumulate_ops` means only the ops used in accumulates for which we expect atomicity.

**Comment by jdinan** [comment]
Wondering whether we might want `units` instead of `bytes` in `accumulate_max_bytes`?  Displacement units are the unit of addressing, but not necessarily of access (since the length of the datatype used can be less than the size of a disp unit).  I suspect the answer to this question is "no", but I wanted to mention in case I'm missing something.

**Comment by pavanbalaji** [comment]
> @pavanbalaji : I guess all the accumulate hints use the term "accumulate" as atomic operation, but not as MPI accumulate call. E.g., `which_accumulate_ops` means only the ops used in accumulates for which we expect atomicity.

I

**Comment by pavanbalaji** [comment]
> @pavanbalaji : I guess all the accumulate hints use the term "accumulate" as atomic operation, but not as MPI accumulate call. E.g., `which_accumulate_ops` means only the ops used in accumulates for which we expect atomicity.

I'm not sure what you are responding to.  But once again, I understand the intent.  I'm only asking about the name of the info key.

**Comment by pavanbalaji** [comment]
> Wondering whether we might want `units` instead of `bytes` in `accumulate_max_bytes`? Displacement units are the unit of addressing, but not necessarily of access (since the length of the datatype used can be less than the size of a disp unit). I suspect the answer to this question is "no", but I wanted to mention in case I'm missing something.

We discussed this in the WG meeting.  People didn't seem to care one way or another, because one can calculate one from the other.

**Comment by minsii** [comment]
> I'm not sure what you are responding to. But once again, I understand the intent. I'm only asking about the name of the info key.

I was trying to figure out whether the info name may cause confusion as you pointed out in a previous comment (see below).

> I guess my point is the name accumulate_derived_dtype == false sounds like "I'll not use derived datatypes at all". However, the intent of the info key is to say "I might use derived datatypes, but I don't expect atomicity for them". I'm just trying to avoid some confusion here. 

As we already use the term "accumulate" as atomic operation in other info keys, there should be no confusion about `accumulate_derived_dtype`.

**Comment by pavanbalaji** [comment]
> > I guess my point is the name accumulate_derived_dtype == false sounds like "I'll not use derived datatypes at all". However, the intent of the info key is to say "I might use derived datatypes, but I don't expect atomicity for them". I'm just trying to avoid some confusion here.
> 
> As we already use the term "accumulate" as atomic operation in other info keys, there should be no confusion about `accumulate_derived_dtype`.

I'd recommend bringing it up in the WG call.  It did manage to confuse you.  😃

**Comment by minsii** [comment]
**Feedback from 2018-12-17 WG call**
At the fastest path where user does not use derived datatypes in accumulates at all (i.e., `accumulate_derived_dtype =false`), `accumulate_max_bytes` seem to be sufficient for both MPICH and OpenMPI to enable the network hardware atomics based optimization.

However, when user uses derived datatypes, the implementation may still choose to use network based routines (e.g., *get-local compute-cswap*,  or *issue chunked data with multiple atomics*). In such case, the information of "chunk size" or "pack size" might be useful for the implementation to make protocol switch for better performance. Can we use the `accumulate_max_bytes` info to provide such information as well ? 

We plan to implement these keys in both MPICH and OpenMPI, thus we will get a better understanding about whether there is any missing information.

@hjelmn please feel free to update the note if anything is missing.

**Comment by jeffhammond** [comment]
I really hope that we don't define these hints such that one contiguous type performs differently from the equivalent arguments without a derived datatype.

**Comment by pavanbalaji** [comment]
I don't think giving `contig` special priority compared with other derived datatypes is a good idea.  Some networks give ability to do simple derived datatypes (such as vectors) in hardware too.  Why are you skipping that?

If you allow some derived datatypes, allow all of them, for example through a list.

**Comment by minsii** [comment]
I think `contig` is the most straightforward one. For other kinds of derived datatype, I am still not sure how to better define "the complexity" of such  a datatype (e.g., does `vector` mean only vector of blocks with basic datatype element, how about `vector of vector`) ?

**Comment by pavanbalaji** [comment]
I find `vector` straightforward too.  Your info key doesn't allow me to add `vector` later.  You essentially hardcoded `contig` into the name.

**Comment by jeffhammond** [comment]
@pavanbalaji `contig` is the only user-defined datatype that maps directly to the usage of built-in ones and can be subjected to self-consistency analysis (i.e. `N*1` should behave like `1*{N}`.  That is why I give it special priority.
